src/Aspire.Hosting.Azure.KeyVault/AzureKeyVaultResourceExtensions.cs

```diff
@@ -103,4 +151,147 @@ public static IResourceBuilder<T> WithRoleAssignments<T>(
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentNullException.ThrowIfNull(parameterResource);
+
+        ValidateSecretName(secretName);
+
+        var secret = new AzureKeyVaultSecretResource(secretName, secretName, builder.Resource, parameterResource);
+        builder.Resource.Secrets.Add(secret);
+
+        return builder.ApplicationBuilder.AddResource(secret);
+    }
+
+    /// <summary>
+    /// Adds a secret to the Azure Key Vault resource with the value from a reference expression.
+    /// </summary>
+    /// <param name="builder">The Azure Key Vault resource builder.</param>
+    /// <param name="secretName">The name of the secret. Must follow Azure Key Vault naming rules.</param>
+    /// <param name="value">The reference expression containing the secret value.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<AzureKeyVaultSecretResource> AddSecret(this IResourceBuilder<AzureKeyVaultResource> builder, string secretName, ReferenceExpression value)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentNullException.ThrowIfNull(value);
+
+        ValidateSecretName(secretName);
+
+        var secret = new AzureKeyVaultSecretResource(secretName, secretName, builder.Resource, value);
+        builder.Resource.Secrets.Add(secret);
+
+        return builder.ApplicationBuilder.AddResource(secret);
+    }
+
+    /// <summary>
+    /// Adds a secret to the Azure Key Vault resource with the value from a parameter resource.
+    /// </summary>
+    /// <param name="builder">The Azure Key Vault resource builder.</param>
+    /// <param name="name">The name of the secret resource.</param>
+    /// <param name="secretName">The name of the secret. Must follow Azure Key Vault naming rules.</param>
+    /// <param name="parameterResource">The parameter resource containing the secret value.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<AzureKeyVaultSecretResource> AddSecret(this IResourceBuilder<AzureKeyVaultResource> builder, [ResourceName] string name, string secretName, IResourceBuilder<ParameterResource> parameterResource)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentNullException.ThrowIfNull(parameterResource);
+
+        return builder.AddSecret(name, secretName, parameterResource.Resource);
+    }
+
+    /// <summary>
+    /// Adds a secret to the Azure Key Vault resource with the value from a parameter resource.
+    /// </summary>
+    /// <param name="builder">The Azure Key Vault resource builder.</param>
+    /// <param name="name">The name of the secret resource.</param>
+    /// <param name="secretName">The name of the secret. Must follow Azure Key Vault naming rules.</param>
+    /// <param name="parameterResource">The parameter resource containing the secret value.</param>
```

Comment:
I'm wondering what the corner cases are going to be here particularly with `azd` where there is a distinct resources then compute deployment cycle.

Reply:
How so?